src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -585,26 +601,18 @@ private HeaderStoreItemInfo CloneHeaderInfo(HeaderDescriptor descriptor, HeaderS
 
             if (descriptor.Parser == null)
             {
-                // We have custom header values. The parsed values are strings.
-                // Custom header values are always stored as string or list of strings.
-                Debug.Assert(sourceInfo.InvalidValue == null, "No invalid values expected for custom headers.");
-                destinationInfo.ParsedValue = CloneStringHeaderInfoValues(sourceInfo.ParsedValue);
+                sourceInfo.AssertContainsNoInvalidValues();
+                destinationInfo.ParsedAndInvalidValues = CloneStringHeaderInfoValues(sourceInfo.ParsedAndInvalidValues);
             }
             else
             {
-                // We have a parser, so we also have to copy invalid values and clone parsed values.
-
-                // Invalid values are always strings. Strings are immutable. So we only have to clone the
-                // collection (if there is one).
-                destinationInfo.InvalidValue = CloneStringHeaderInfoValues(sourceInfo.InvalidValue);
-
                 // Now clone and add parsed values (if any).
-                if (sourceInfo.ParsedValue != null)
+                if (sourceInfo.ParsedAndInvalidValues != null)
                 {
-                    List<object>? sourceValues = sourceInfo.ParsedValue as List<object>;
+                    List<object>? sourceValues = sourceInfo.ParsedAndInvalidValues as List<object>;
                     if (sourceValues == null)
                     {
-                        CloneAndAddValue(destinationInfo, sourceInfo.ParsedValue);
+                        CloneAndAddValue(destinationInfo, sourceInfo.ParsedAndInvalidValues);
```

Comment:
What happens with InvalidValue here? Records are immutable, right? And they don't implement ICloneable, right?

Seems like this is worth a comment in CloneAndAddValue below, possibly an assert as well.

Reply:
@feiyun0112 this was marked as resolved, but I don't see any related changes

Reply:
committed